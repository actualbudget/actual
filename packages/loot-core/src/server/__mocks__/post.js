export {
  handleRequest as post,
  handleRequestBinary as postBinary,
} from '../tests/mockSyncServer';

export const get = function (url) {
  throw new Error('get unimplemented');
};
