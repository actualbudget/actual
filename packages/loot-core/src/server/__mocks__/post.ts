// @ts-strict-ignore
export {
  handleRequest as post,
  handleRequestBinary as postBinary,
} from '../tests/mockSyncServer';

export const get = function () {
  throw new Error('get unimplemented');
};
