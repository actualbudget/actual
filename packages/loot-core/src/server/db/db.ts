// @ts-strict-ignore
import {
  makeClock,
  setClock,
  serializeClock,
  deserializeClock,
  makeClientId,
  Timestamp,
} from '@actual-app/crdt';
import { Database } from '@jlongster/sql.js';
import { LRUCache } from 'lru-cache';
import { v4 as uuidv4 } from 'uuid';

import * as fs from '../../platform/server/fs';
import * as sqlite from '../../platform/server/sqlite';
import {
  schema,
  schemaConfig,
  convertForInsert,
  convertForUpdate,
  convertFromSelect,
} from '../aql';
import { sendMessages } from '../sync';

import { DbClockMessage } from './types';

export * from './types';

export { toDateRepr, fromDateRepr } from '../models';

let dbPath: string | null = null;
let db: Database | null = null;

// Util

export function getDatabasePath() {
  return dbPath;
}

export async function openDatabase(id?: string) {
  if (db) {
    await sqlite.closeDatabase(db);
  }

  dbPath = fs.join(fs.getBudgetDir(id), 'db.sqlite');
  setDatabase(await sqlite.openDatabase(dbPath));

  // await execQuery('PRAGMA journal_mode = WAL');
}

export async function closeDatabase() {
  if (db) {
    await sqlite.closeDatabase(db);
    setDatabase(null);
  }
}

export function setDatabase(db_: Database) {
  db = db_;
  resetQueryCache();
}

export function getDatabase() {
  return db;
}

export async function loadClock() {
  const row = await first<DbClockMessage>('SELECT * FROM messages_clock');
  if (row) {
    const clock = deserializeClock(row.clock);
    setClock(clock);
  } else {
    // No clock exists yet (first run of the app), so create a default
    // one.
    const timestamp = new Timestamp(0, 0, makeClientId());
    const clock = makeClock(timestamp);
    setClock(clock);

    await runQuery('INSERT INTO messages_clock (id, clock) VALUES (?, ?)', [
      1,
      serializeClock(clock),
    ]);
  }
}

// Functions
export function runQuery(
  sql: string,
  params?: Array<string | number>,
  fetchAll?: false,
): { changes: unknown };

export function runQuery<T>(
  sql: string,
  params: Array<string | number> | undefined,
  fetchAll: true,
): T[];

export function runQuery<T>(
  sql: string,
  params: (string | number)[],
  fetchAll: boolean,
) {
  if (fetchAll) {
    return sqlite.runQuery<T>(db, sql, params, true);
  } else {
    return sqlite.runQuery(db, sql, params, false);
  }
}

export function execQuery(sql: string) {
  sqlite.execQuery(db, sql);
}

// This manages an LRU cache of prepared query statements. This is
// only needed in hot spots when you are running lots of queries.
let _queryCache = new LRUCache<string, string>({ max: 100 });
export function cache(sql: string) {
  const cached = _queryCache.get(sql);
  if (cached) {
    return cached;
  }

  const prepared = sqlite.prepare(db, sql);
  _queryCache.set(sql, prepared);
  return prepared;
}

function resetQueryCache() {
  _queryCache = new LRUCache<string, string>({ max: 100 });
}

export function transaction(fn: () => void) {
  return sqlite.transaction(db, fn);
}

export function asyncTransaction(fn: () => Promise<void>) {
  return sqlite.asyncTransaction(db, fn);
}

// This function is marked as async because `runQuery` is no longer
// async. We return a promise here until we've audited all the code to
// make sure nothing calls `.then` on this.
export async function all<T>(sql: string, params?: (string | number)[]) {
  return runQuery<T>(sql, params, true);
}

export async function first<T>(sql, params?: (string | number)[]) {
  const arr = await runQuery<T>(sql, params, true);
  return arr.length === 0 ? null : arr[0];
}

// The underlying sql system is now sync, but we can't update `first` yet
// without auditing all uses of it
export function firstSync<T>(sql, params?: (string | number)[]) {
  const arr = runQuery<T>(sql, params, true);
  return arr.length === 0 ? null : arr[0];
}

// This function is marked as async because `runQuery` is no longer
// async. We return a promise here until we've audited all the code to
// make sure nothing calls `.then` on this.
export async function run(sql, params?: (string | number)[]) {
  return runQuery(sql, params);
}

export async function select(table, id) {
  const rows = await runQuery(
    'SELECT * FROM ' + table + ' WHERE id = ?',
    [id],
    true,
  );
  // TODO: In the next phase, we will make this function generic
  // and pass the type of the return type to `runQuery`.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return rows[0] as any;
}

export async function update(table, params) {
  const fields = Object.keys(params).filter(k => k !== 'id');

  if (params.id == null) {
    throw new Error('update: id is required');
  }

  await sendMessages(
    fields.map(k => {
      return {
        dataset: table,
        row: params.id,
        column: k,
        value: params[k],
        timestamp: Timestamp.send(),
      };
    }),
  );
}

export async function insertWithUUID(table, row) {
  if (!row.id) {
    row = { ...row, id: uuidv4() };
  }

  await insert(table, row);

  // We can't rely on the return value of insert because if the
  // primary key is text, sqlite returns the internal row id which we
  // don't care about. We want to return the generated UUID.
  return row.id;
}

export async function insert(table, row) {
  const fields = Object.keys(row).filter(k => k !== 'id');

  if (row.id == null) {
    throw new Error('insert: id is required');
  }

  await sendMessages(
    fields.map(k => {
      return {
        dataset: table,
        row: row.id,
        column: k,
        value: row[k],
        timestamp: Timestamp.send(),
      };
    }),
  );
}

export async function delete_(table, id) {
  await sendMessages([
    {
      dataset: table,
      row: id,
      column: 'tombstone',
      value: 1,
      timestamp: Timestamp.send(),
    },
  ]);
}

export async function deleteAll(table: string) {
  const rows = await all<{ id: string }>(`
    SELECT id FROM ${table} WHERE tombstone = 0
  `);
  await Promise.all(rows.map(({ id }) => delete_(table, id)));
}

export async function selectWithSchema(table, sql, params) {
  const rows = await runQuery(sql, params, true);
  const convertedRows = rows
    .map(row => convertFromSelect(schema, schemaConfig, table, row))
    .filter(Boolean);
  // TODO: Make convertFromSelect generic so we don't need this cast
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return convertedRows as any[];
}

export async function selectFirstWithSchema(table, sql, params) {
  const rows = await selectWithSchema(table, sql, params);
  return rows.length > 0 ? rows[0] : null;
}

export function insertWithSchema(table, row) {
  // Even though `insertWithUUID` does this, we need to do it here so
  // the schema validation passes
  if (!row.id) {
    row = { ...row, id: uuidv4() };
  }

  return insertWithUUID(
    table,
    convertForInsert(schema, schemaConfig, table, row),
  );
}

export function updateWithSchema(table, fields) {
  return update(table, convertForUpdate(schema, schemaConfig, table, fields));
}
