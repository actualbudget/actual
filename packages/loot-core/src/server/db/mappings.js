import { addSyncListener } from '../sync/index';

import * as db from './index';

// This file keeps all the mappings in memory so we can access it
// synchronously. This is primarily used in the rules system, but
// there may be other uses in the future. You don't need to worry
// about this generally; if you are querying transactions, ids are
// transparently mapped for you. But if you are building something
// that stores ids and later uses them, you need to remember to map
// the ids.
//
// IMPORTANT: `loadMappings` must be called first before other modules
// that listen for sync changes. This must be the first sync listener
// to run in case other listeners use this mapping table; otherwise
// they might see stale mappings.

let allMappings;
let unlistenSync;

export async function loadMappings() {
  // The mappings are separated into tables specific to the type of
  // data. But you know, we really could keep a global mapping table.
  let categories = (await db.all('SELECT * FROM category_mapping')).map(r => [
    r.id,
    r.transferId,
  ]);
  let payees = (await db.all('SELECT * FROM payee_mapping')).map(r => [
    r.id,
    r.targetId,
  ]);

  // All ids are unique, so we can just keep a global table of mappings
  allMappings = new Map(categories.concat(payees));

  if (unlistenSync) {
    unlistenSync();
  }
  unlistenSync = addSyncListener(onApplySync);
}

function onApplySync(oldValues, newValues) {
  newValues.forEach((items, table) => {
    if (table.indexOf('mapping') !== -1) {
      let field = table === 'category_mapping' ? 'transferId' : 'targetId';

      items.forEach(newValue => {
        allMappings.set(newValue.id, newValue[field]);
      });
    }
  });
}

export function getMappings() {
  return allMappings;
}

export function getMapping(id) {
  return allMappings.get(id) || null;
}
