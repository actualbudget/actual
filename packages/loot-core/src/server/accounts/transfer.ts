import * as db from '../db';

async function getPayee(acct) {
  return db.first('SELECT * FROM payees WHERE transfer_acct = ?', [acct]);
}

async function getTransferredAccount(transaction) {
  if (transaction.payee) {
    let { transfer_acct } = await db.first(
      'SELECT id, transfer_acct FROM v_payees WHERE id = ?',
      [transaction.payee],
    );
    return transfer_acct;
  }
  return null;
}

async function clearCategory(transaction, transferAcct) {
  const { offbudget: fromOffBudget } = await db.first(
    'SELECT offbudget FROM accounts WHERE id = ?',
    [transaction.account],
  );
  const { offbudget: toOffBudget } = await db.first(
    'SELECT offbudget FROM accounts WHERE id = ?',
    [transferAcct],
  );

  // We should clear the category to make sure it's not being
  // accounted for in the budget, unless it should be in the case of
  // transferring from an on-budget to off-budget account
  if (fromOffBudget === toOffBudget) {
    await db.updateTransaction({ id: transaction.id, category: null });
    return true;
  }
  return false;
}

export async function addTransfer(transaction, transferredAccount) {
  if (transaction.is_parent) {
    // For split transactions, we should create transfers using child transactions.
    // This is to ensure that the amounts received by the transferred account
    // reflects the amounts in the child transactions and not the parent transaction
    // amount which is the total amount.
    return null;
  }

  let { id: fromPayee } = await db.first(
    'SELECT id FROM payees WHERE transfer_acct = ?',
    [transaction.account],
  );

  // We need to enforce certain constraints with child transaction transfers
  if (transaction.parent_id) {
    let row = await db.first(
      `
        SELECT p.id, p.transfer_acct FROM v_transactions t
        LEFT JOIN payees p ON p.id = t.payee
        WHERE t.id = ?
      `,
      [transaction.parent_id],
    );

    if (row.transfer_acct) {
      if (row.id !== transaction.payee) {
        // This child transaction is trying to use a transfer payee,
        // but the parent is already using a different transfer payee.
        // This is not allowed, so not only do we do nothing, we clear
        // the payee of the child transaction to make it clear
        await db.updateTransaction({ id: transaction.id, payee: null });
        return { id: transaction.id, payee: null };
      }
    }
  }

  const id = await db.insertTransaction({
    account: transferredAccount,
    amount: -transaction.amount,
    payee: fromPayee,
    date: transaction.date,
    transfer_id: transaction.id,
    notes: transaction.notes || null,
    cleared: false,
  });

  await db.updateTransaction({ id: transaction.id, transfer_id: id });
  const categoryCleared = await clearCategory(transaction, transferredAccount);

  return {
    id: transaction.id,
    transfer_id: id,
    ...(categoryCleared ? { category: null } : {}),
  };
}

export async function removeTransfer(transaction) {
  let transferTrans = await db.getTransaction(transaction.transfer_id);

  // Perform operations on the transfer transaction only
  // if it is found. For example: when users delete both
  // (in & out) transfer transactions at the same time -
  // transfer transaction will not be found.
  if (transferTrans) {
    if (transferTrans.is_child) {
      // If it's a child transaction, we don't delete it because that
      // would invalidate the whole split transaction. Instead of turn
      // it into a normal transaction
      await db.updateTransaction({
        id: transaction.transfer_id,
        transfer_id: null,
        payee: null,
      });
    } else {
      await db.deleteTransaction({ id: transaction.transfer_id });
    }
  }
  await db.updateTransaction({ id: transaction.id, transfer_id: null });
  return { id: transaction.id, transfer_id: null };
}

export async function updateTransfer(transaction, transferredAccount) {
  let payee = await getPayee(transaction.account);

  await db.updateTransaction({
    id: transaction.transfer_id,
    account: transferredAccount,
    // Make sure to update the payee on the other side in case the
    // user moved this transaction into another account
    payee: payee.id,
    date: transaction.date,
    notes: transaction.notes,
    amount: -transaction.amount,
  });

  const categoryCleared = await clearCategory(transaction, transferredAccount);
  if (categoryCleared) {
    return { id: transaction.id, category: null };
  }
}

export async function onInsert(transaction) {
  let transferredAccount = await getTransferredAccount(transaction);

  if (transferredAccount) {
    return addTransfer(transaction, transferredAccount);
  }
}

export async function onDelete(transaction) {
  if (transaction.transfer_id) {
    await removeTransfer(transaction);
  }
}

export async function onUpdate(transaction) {
  const transferredAccount = await getTransferredAccount(transaction);

  if (transferredAccount && !transaction.transfer_id) {
    return addTransfer(transaction, transferredAccount);
  }

  if (!transferredAccount && transaction.transfer_id) {
    return removeTransfer(transaction);
  }

  if (transferredAccount && transaction.transfer_id) {
    return updateTransfer(transaction, transferredAccount);
  }
}
