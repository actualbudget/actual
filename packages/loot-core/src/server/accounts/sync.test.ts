import snapshotDiff from 'snapshot-diff';

import * as monthUtils from '../../shared/months';
import * as db from '../db';
import { loadMappings } from '../db/mappings';
import { post } from '../post';
import { getServer } from '../server-config';
import * as mockSyncServer from '../tests/mockSyncServer';

import {
  syncAccount,
  reconcileTransactions,
  addTransactions,
  fromPlaid,
} from './sync';
import { loadRules, insertRule } from './transaction-rules';
import * as transfer from './transfer';

const papaJohns = 'Papa Johns east side';
const lowes = 'Lowe’s Store';

jest.mock('../../shared/months', () => ({
  ...jest.requireActual('../../shared/months'),
  currentDay: jest.fn(),
  currentMonth: jest.fn(),
}));

beforeEach(async () => {
  mockSyncServer.reset();
  jest.resetAllMocks();
  (monthUtils.currentDay as jest.Mock).mockReturnValue('2017-10-15');
  (monthUtils.currentMonth as jest.Mock).mockReturnValue('2017-10');
  await global.emptyDatabase()();
  await loadMappings();
  await loadRules();
});

function getAllTransactions() {
  return db.all(
    `SELECT t.*, p.name as payee_name
       FROM v_transactions_internal t
       LEFT JOIN payees p ON p.id = t.payee
       ORDER BY date DESC, amount DESC, id
     `,
  );
}

function expectSnapshotWithDiffer(initialValue) {
  let currentValue = initialValue;
  expect(initialValue).toMatchSnapshot();
  return {
    expectToMatchDiff: value => {
      expect(snapshotDiff(currentValue, value)).toMatchSnapshot();
      currentValue = value;
    },
  };
}

function prepMockTransactions() {
  let mockTransactions;
  mockSyncServer.filterMockData(data => {
    const account_id = data.accounts[0].account_id;
    const transactions = data.transactions[account_id].filter(t => !t.pending);

    mockTransactions = [
      ...transactions.filter(t => t.date <= '2017-10-15'),
      ...transactions.filter(t => t.date === '2017-10-16').slice(0, 1),
      ...transactions.filter(t => t.date === '2017-10-17').slice(0, 3),
    ];

    return {
      accounts: data.accounts,
      transactions: { [account_id]: mockTransactions },
    };
  });
  return mockTransactions;
}

async function prepareDatabase() {
  await db.insertCategoryGroup({ id: 'group1', name: 'group1', is_income: 1 });
  await db.insertCategory({
    name: 'income',
    cat_group: 'group1',
    is_income: 1,
  });

  const { accounts } = await post(getServer().PLAID_SERVER + '/accounts', {
    client_id: '',
    group_id: '',
    item_id: '1',
  });
  const acct = accounts[0];

  const id = await db.insertAccount({
    id: 'one',
    account_id: acct.account_id,
    name: acct.official_name,
    balance_current: acct.balances.current,
  });
  await db.insertPayee({
    id: 'transfer-' + id,
    name: '',
    transfer_acct: id,
  });

  return { id, account_id: acct.account_id };
}

async function getAllPayees() {
  return (await db.getPayees()).filter(p => p.transfer_acct == null);
}

describe('Account sync', () => {
  test('reconcile creates payees correctly', async () => {
    let { id } = await prepareDatabase();

    let payees = await getAllPayees();
    expect(payees.length).toBe(0);

    await reconcileTransactions(id, [
      { date: '2020-01-02', payee_name: 'bakkerij', amount: 4133 },
      { date: '2020-01-03', payee_name: 'kroger', amount: 5000 },
    ]);

    payees = await getAllPayees();
    expect(payees.length).toBe(2);

    let transactions = await getAllTransactions();
    expect(transactions.length).toBe(2);
    expect(transactions.find(t => t.amount === 4133).payee).toBe(
      payees.find(p => p.name === 'Bakkerij').id,
    );
    expect(transactions.find(t => t.amount === 5000).payee).toBe(
      payees.find(p => p.name === 'Kroger').id,
    );
  });

  test('reconcile matches single transaction', async () => {
    let mockTransactions = prepMockTransactions();
    const { id, account_id } = await prepareDatabase();

    await syncAccount('userId', 'userKey', id, account_id, 'bank');

    // The payee can be anything, all that matters is the amount is the same
    let mockTransaction = mockTransactions.find(t => t.date === '2017-10-17');
    mockTransaction.amount = 29.47;

    let payeeId = await db.insertPayee({ name: 'macy' });
    await db.insertTransaction({
      id: 'one',
      account: id,
      amount: -2947,
      date: '2017-10-15',
      payee: payeeId,
    });

    let { added, updated } = await reconcileTransactions(
      id,
      mockTransactions.filter(t => t.date >= '2017-10-15').map(fromPlaid),
    );

    expect(added.length).toBe(3);
    expect(updated.length).toBe(1);

    let transactions = await getAllTransactions();
    let transaction = transactions.find(t => t.amount === -2947);
    expect(transaction.id).toBe(updated[0]);

    // The payee has not been updated - it's still the payee that the original transaction had
    let payees = await getAllPayees();
    expect(payees.length).toBe(18);
    expect(transaction.payee).toBe(payeeId);
  });

  test('reconcile matches multiple transactions', async () => {
    let mockTransactions = prepMockTransactions();
    const { id, account_id } = await prepareDatabase();

    await syncAccount('userId', 'userKey', id, account_id, 'bank');

    // These should all match, but note that the one with the payee
    // `macy` should match with the imported one with the same payee
    // name. This should happen even though other transactions with
    // the same amount are imported first, i.e. high fidelity matches
    // always win
    let mocked = mockTransactions.filter(t => t.date === '2017-10-17');
    mocked[0].name = papaJohns;
    mocked[0].amount = 29.47;
    mocked[1].name = 'Lowe’s Store';
    mocked[1].amount = 29.47;
    mocked[2].name = 'macy';
    mocked[2].amount = 29.47;

    // Make sure that it macy is correctly matched from a different
    // day first, and then the other two are matched based on amount.
    // And it should never match the same transactions twice
    await db.insertTransaction({
      id: 'one',
      account: id,
      amount: -2947,
      date: '2017-10-15',
      payee: await db.insertPayee({ name: 'papa johns' }),
    });
    await db.insertTransaction({
      id: 'two',
      account: id,
      amount: -2947,
      date: '2017-10-17',
      payee: await db.insertPayee({ name: 'lowes' }),
    });
    await db.insertTransaction({
      id: 'three',
      account: id,
      amount: -2947,
      date: '2017-10-17',
      payee: await db.insertPayee({ name: 'macy' }),
    });

    let { added, updated } = await reconcileTransactions(
      id,
      mockTransactions.filter(t => t.date >= '2017-10-15').map(fromPlaid),
    );

    let transactions = await getAllTransactions();
    expect(updated.length).toBe(3);
    expect(added.length).toBe(1);

    expect(transactions.find(t => t.id === 'one').imported_id).toBe(
      mocked[1].transaction_id,
    );
    expect(transactions.find(t => t.id === 'two').imported_id).toBe(
      mocked[0].transaction_id,
    );
    expect(transactions.find(t => t.id === 'three').imported_id).toBe(
      mocked[2].transaction_id,
    );
  });

  test('reconcile matches multiple transactions (imported_id wins)', async () => {
    let mockTransactions = prepMockTransactions();
    const { id, account_id } = await prepareDatabase();

    await syncAccount('userId', 'userKey', id, account_id, 'bank');

    let mocked = mockTransactions.filter(t => t.date === '2017-10-17');
    mocked[0].name = papaJohns;
    mocked[0].amount = 29.47;
    mocked[1].name = lowes;
    mocked[1].amount = 29.47;
    mocked[1].transaction_id = 'imported1';

    // Technically, the amount doesn't even matter. The
    // imported_id will always match no matter what
    await db.insertTransaction({
      id: 'one',
      account: id,
      amount: -3000,
      date: '2017-10-15',
      imported_id: 'imported1',
      payee: await db.insertPayee({ name: 'papa johns' }),
    });
    await db.insertTransaction({
      id: 'two',
      account: id,
      amount: -2947,
      date: '2017-10-17',
      payee: await db.insertPayee({ name: 'lowes' }),
    });

    let { added, updated } = await reconcileTransactions(
      id,
      mockTransactions.filter(t => t.date >= '2017-10-15').map(fromPlaid),
    );

    let transactions = await getAllTransactions();
    expect(updated).toEqual(['two', 'one']);
    expect(added.length).toBe(2);

    // Make sure lowes, which has the imported_id, is the one that
    // got matched with the same imported_id
    expect(transactions.find(t => t.id === 'one').imported_payee).toBe(lowes);
  });

  test('import never matches existing with financial ids', async () => {
    let mockTransactions = prepMockTransactions();
    const { id, account_id } = await prepareDatabase();

    await syncAccount('userId', 'userKey', id, account_id, 'bank');
    let differ = expectSnapshotWithDiffer(await getAllTransactions());

    mockTransactions = mockTransactions.filter(t => t.date === '2017-10-17');
    mockTransactions[0].name = 'foo';
    mockTransactions[0].amount = 29.47;
    mockTransactions[1].name = 'bar';
    mockTransactions[1].amount = 29.47;

    // Make sure, no matter what, it never tries to match with an
    // existing transaction that already has a financial id
    await db.insertTransaction({
      id: 'one',
      account: id,
      amount: -2947,
      date: '2017-10-15',
      payee: await db.insertPayee({ name: 'foo' }),
      imported_id: 'trans1',
    });

    await db.insertTransaction({
      id: 'two',
      account: id,
      amount: -2947,
      date: '2017-10-15',
      payee: await db.insertPayee({ name: 'bar' }),
      imported_id: 'trans2',
    });

    differ.expectToMatchDiff(await getAllTransactions());

    (monthUtils.currentDay as jest.Mock).mockReturnValue('2017-10-17');
    await syncAccount('userId', 'userKey', id, account_id, 'bank');

    differ.expectToMatchDiff(await getAllTransactions());
  });

  test('import updates transfers when matched', async () => {
    let mockTransactions = prepMockTransactions();
    const { id, account_id } = await prepareDatabase();
    await db.insertAccount({ id: 'two', name: 'two' });
    await db.insertPayee({
      id: 'transfer-two',
      name: '',
      transfer_acct: 'two',
    });

    await syncAccount('userId', 'userKey', id, account_id, 'bank');
    let differ = expectSnapshotWithDiffer(await getAllTransactions());

    const mockTransaction = mockTransactions.find(t => t.date === '2017-10-17');
    mockTransaction.name = '#001 fenn st Macy’s 33333 EMX';
    mockTransaction.amount = 29.48;

    const transactionId = await db.insertTransaction({
      id: 'one',
      account: 'two',
      amount: 2948,
      date: '2017-10-15',
      payee: 'transfer-' + id,
    });
    await transfer.onInsert(await db.getTransaction(transactionId));

    differ.expectToMatchDiff(await getAllTransactions());

    (monthUtils.currentDay as jest.Mock).mockReturnValue('2017-10-17');
    await syncAccount('userId', 'userKey', id, account_id, 'bank');

    // Don't use `differ.expectToMatchDiff` because there's too many
    // changes that look too confusing
    expect(await getAllTransactions()).toMatchSnapshot();
  });

  test('reconcile handles transactions with undefined fields', async () => {
    const { id: acctId } = await prepareDatabase();

    await db.insertTransaction({
      id: 'one',
      account: acctId,
      amount: 2948,
      date: '2020-01-01',
    });

    await reconcileTransactions(acctId, [
      { date: '2020-01-02' },
      { date: '2020-01-01', amount: 2948 },
    ]);

    let transactions = await getAllTransactions();
    expect(transactions.length).toBe(2);
    expect(transactions).toMatchSnapshot();

    // No payees should be created
    let payees = await getAllPayees();
    expect(payees.length).toBe(0);

    // Make _at least_ the date is required
    await expect(reconcileTransactions(acctId, [{}])).rejects.toThrow(
      /`date` is required/,
    );
  });

  test('reconcile run rules with inferred payee', async () => {
    const { id: acctId } = await prepareDatabase();
    await db.insertCategoryGroup({
      id: 'group2',
      name: 'group2',
    });
    let catId = await db.insertCategory({
      name: 'Food',
      cat_group: 'group2',
    });

    let payeeId = await db.insertPayee({ name: 'bakkerij' });

    await insertRule({
      stage: null,
      conditionsOp: 'and',
      conditions: [{ op: 'is', field: 'payee', value: payeeId }],
      actions: [{ op: 'set', field: 'category', value: catId }],
    });

    await reconcileTransactions(acctId, [
      { date: '2020-01-02', payee_name: 'Bakkerij', amount: 4133 },
    ]);

    let transactions = await getAllTransactions();
    // Even though the payee was inferred from the string name (no
    // renaming rules ran), it should match the above rule and set the
    // category
    expect(transactions.length).toBe(1);
    expect(transactions[0].payee).toBe(payeeId);
    expect(transactions[0].category).toBe(catId);

    // It also should not have created a payee
    let payees = await getAllPayees();
    expect(payees.length).toBe(1);
    expect(payees[0].id).toBe(payeeId);
  });

  test('reconcile avoids creating blank payees', async () => {
    const { id: acctId } = await prepareDatabase();

    await reconcileTransactions(acctId, [
      { date: '2020-01-02', payee_name: '     ', amount: 4133 },
    ]);

    let transactions = await getAllTransactions();
    // Even though the payee was inferred from the string name (no
    // renaming rules ran), it should match the above rule and set the
    // category
    expect(transactions.length).toBe(1);
    expect(transactions[0].payee).toBe(null);
    expect(transactions[0].amount).toBe(4133);
    expect(transactions[0].date).toBe(20200102);

    // It also should not have created a payee
    let payees = await getAllPayees();
    expect(payees.length).toBe(0);
  });

  test('reconcile run rules dont create unnecessary payees', async () => {
    const { id: acctId } = await prepareDatabase();

    let payeeId = await db.insertPayee({ name: 'bakkerij-renamed' });

    await insertRule({
      stage: null,
      conditionsOp: 'and',
      conditions: [{ op: 'is', field: 'imported_payee', value: 'Bakkerij' }],
      actions: [{ op: 'set', field: 'payee', value: payeeId }],
    });

    await reconcileTransactions(acctId, [
      { date: '2020-01-02', payee_name: 'bakkerij', amount: 4133 },
    ]);

    let payees = await getAllPayees();
    expect(payees.length).toBe(1);
    expect(payees[0].id).toBe(payeeId);

    let transactions = await getAllTransactions();
    expect(transactions.length).toBe(1);
    expect(transactions[0].payee).toBe(payeeId);
  });

  let testMapped = version => {
    test(`reconcile matches unmapped and mapped payees (${version})`, async () => {
      const { id: acctId } = await prepareDatabase();

      if (version === 'v1') {
        // This is quite complicated, but important to test. If a payee is
        // merged with another, a rule sets the payee of a transaction to
        // the updated one, make sure it still matches an existing
        // transaction that points to the old merged payee
      } else if (version === 'v2') {
        // This is similar to v1, but inverted: make sure that
        // if a rule sets the payee to an *old* payee, that it still
        // matches to a transaction with the new payee that it was merged
        // to
      }

      let payeeId1 = await db.insertPayee({ name: 'bakkerij2' });
      let payeeId2 = await db.insertPayee({ name: 'bakkerij-renamed' });

      // Insert a rule *before* payees are merged. Not that v2 would
      // fail if we inserted this rule after, because the rule would
      // set to an *old* payee but the matching would take place on a
      // *new* payee. But that's ok - it would fallback to matching
      // amount anyway, so while it loses some fidelity, it's an edge
      // case that we don't need to worry much about because the user
      // shouldn't be able able to create rules for a merged payee.
      // Unless they sync in a rule...
      await insertRule({
        stage: null,
        conditionsOp: 'and',
        conditions: [{ op: 'is', field: 'imported_payee', value: 'Bakkerij' }],
        actions: [{ op: 'set', field: 'payee', value: payeeId2 }],
      });

      if (version === 'v1') {
        await db.mergePayees(payeeId2, [payeeId1]);
      } else if (version === 'v2') {
        await db.mergePayees(payeeId1, [payeeId2]);
      }

      await db.insertTransaction({
        id: 'one',
        account: acctId,
        amount: -2947,
        date: '2017-10-15',
        payee: payeeId1,
      });
      // It will try to match to this one first, make sure it matches
      // the above transaction though
      await db.insertTransaction({
        id: 'two',
        account: acctId,
        amount: -2947,
        date: '2017-10-17',
        payee: null,
      });

      let { updated } = await reconcileTransactions(acctId, [
        {
          date: '2017-10-17',
          payee_name: 'bakkerij',
          amount: -2947,
          imported_id: 'imported1',
        },
      ]);

      let payees = await getAllPayees();
      expect(payees.length).toBe(1);
      expect(payees[0].id).toBe(version === 'v1' ? payeeId2 : payeeId1);

      expect(updated.length).toBe(1);
      expect(updated[0]).toBe('one');

      let transactions = await getAllTransactions();
      expect(transactions.length).toBe(2);
      expect(transactions.find(t => t.id === 'one').imported_id).toBe(
        'imported1',
      );
    });
  };

  testMapped('v1');
  testMapped('v2');

  test('addTransactions simply adds transactions', async () => {
    const { id: acctId } = await prepareDatabase();

    let payeeId = await db.insertPayee({ name: 'bakkerij-renamed' });

    // Make sure it still runs rules
    await insertRule({
      stage: null,
      conditionsOp: 'and',
      conditions: [{ op: 'is', field: 'imported_payee', value: 'Bakkerij' }],
      actions: [{ op: 'set', field: 'payee', value: payeeId }],
    });

    let transactions = [
      {
        date: '2017-10-17',
        payee_name: 'BAKKerij',
        amount: -2947,
      },
      {
        date: '2017-10-18',
        payee_name: 'bakkERIj2',
        amount: -2947,
      },
      {
        date: '2017-10-19',
        payee_name: 'bakkerij3',
        amount: -2947,
      },
      {
        date: '2017-10-20',
        payee_name: 'BakkeriJ3',
        amount: -2947,
      },
    ];

    let added = await addTransactions(acctId, transactions);
    expect(added.length).toBe(transactions.length);

    let payees = await getAllPayees();
    expect(payees.length).toBe(3);

    let getName = id => payees.find(p => p.id === id).name;

    let allTransactions = await getAllTransactions();
    expect(allTransactions.length).toBe(4);
    expect(allTransactions.map(t => getName(t.payee))).toEqual([
      'bakkerij3',
      'bakkerij3',
      'bakkERIj2',
      'bakkerij-renamed',
    ]);
  });

  test('imports transactions for current day and adds latest', async () => {
    const { id, account_id } = await prepareDatabase();

    expect((await getAllTransactions()).length).toBe(0);
    await syncAccount('userId', 'userKey', id, account_id, 'bank');
    expect(await getAllTransactions()).toMatchSnapshot();

    (monthUtils.currentDay as jest.Mock).mockReturnValue('2017-10-17');

    await syncAccount('userId', 'userKey', id, account_id, 'bank');
    expect(await getAllTransactions()).toMatchSnapshot();
  });
});
