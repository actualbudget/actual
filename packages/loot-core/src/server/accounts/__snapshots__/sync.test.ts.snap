// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account sync reconcile handles transactions with undefined fields 1`] = `
Array [
  Object {
    "account": "one",
    "amount": 0,
    "category": null,
    "cleared": 1,
    "date": 20200102,
    "error": null,
    "id": "id3",
    "imported_id": null,
    "imported_payee": null,
    "is_child": 0,
    "is_parent": 0,
    "notes": null,
    "parent_id": null,
    "payee": null,
    "payee_name": null,
    "reconciled": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transfer_id": null,
  },
  Object {
    "account": "one",
    "amount": 2948,
    "category": null,
    "cleared": 1,
    "date": 20200101,
    "error": null,
    "id": "one",
    "imported_id": null,
    "imported_payee": null,
    "is_child": 0,
    "is_parent": 0,
    "notes": null,
    "parent_id": null,
    "payee": null,
    "payee_name": null,
    "reconciled": 0,
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transfer_id": null,
  },
]
`;
